File: L6/EmployeeManagement (JavaFX)/src/main/java/com/example/employeemanagement/EmployeeManagement.java
package com.example.employeemanagement;

import javafx.application.Application;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.collections.transformation.FilteredList;
import javafx.collections.transformation.SortedList;
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.HBox;
import javafx.scene.layout.VBox;
import javafx.stage.FileChooser;
import javafx.stage.Stage;
import java.io.*;
import java.util.Comparator;
import java.util.ArrayList;
import java.util.List;

public class EmployeeManagement extends Application {
    private ObservableList<Employee> employees = FXCollections.observableArrayList();
    private FilteredList<Employee> filteredEmployees = new FilteredList<>(employees);
    private SortedList<Employee> sortedEmployees = new SortedList<>(filteredEmployees);
    private boolean sortAscending = true;

    @Override
    public void start(Stage primaryStage) {
        // Create UI elements
        ListView<Employee> employeeListView = new ListView<>(sortedEmployees);
        TextField searchField = new TextField();
        searchField.setPromptText("Search by Last Name");
        searchField.textProperty().addListener((observable, oldValue, newValue) -> {
            // Update filtered employees based on search text
            filteredEmployees.setPredicate(employee -> {
                if (newValue == null || newValue.isEmpty()) {
                    return true;
                }
                String lowerCaseFilter = newValue.toLowerCase();
                return employee.getLastName().toLowerCase().contains(lowerCaseFilter);
            });
        });;
        TextField lastNameField = new TextField();
        lastNameField.setPromptText("Last Name");
        TextField positionField = new TextField();
        positionField.setPromptText("Position");
        TextField departmentField = new TextField();
        departmentField.setPromptText("Department");
        TextField badgeNumberField = new TextField();
        badgeNumberField.setPromptText("Badge Number");
        CheckBox isPresentCheckBox = new CheckBox("Is Present");
        TextField lastEntryExitTimeField = new TextField();
        lastEntryExitTimeField.setPromptText("Last Entry/Exit Time");

        Button addButton = new Button("Add Employee");
        addButton.setOnAction(event -> {
            // Add employee to list
            String lastName = lastNameField.getText();
            String position = positionField.getText();
            String department = departmentField.getText();
            int badgeNumber = Integer.parseInt(badgeNumberField.getText());
            boolean isPresent = isPresentCheckBox.isSelected();
            String lastEntryExitTime = lastEntryExitTimeField.getText();

            Employee employee = new Employee(lastName, position, department, badgeNumber, isPresent, lastEntryExitTime);
            employees.add(employee);

            // Clear input fields
            lastNameField.clear();
            positionField.clear();
            departmentField.clear();
            badgeNumberField.clear();
            isPresentCheckBox.setSelected(false);
            lastEntryExitTimeField.clear();
        });
        
        ContextMenu contextMenu = new ContextMenu();
        MenuItem removeItem = new MenuItem("Remove");
        removeItem.setOnAction(e -> {
            // Remove selected employee from list
            Employee selectedEmployee = employeeListView.getSelectionModel().getSelectedItem();
            employees.remove(selectedEmployee);
        });
        contextMenu.getItems().add(removeItem);
        employeeListView.setContextMenu(contextMenu);

        Button sortButton = new Button("Sort by Last Name");
        sortButton.setOnAction(event -> {
            // Sort employees by last name
            Comparator<Employee> comparator;
            if (sortAscending) {
                comparator = Comparator.comparing(Employee::getLastName);
                sortAscending = false;
                sortButton.setText("Sort by Last Name (Descending)");
            } else {
                comparator = Comparator.comparing(Employee::getLastName).reversed();
                sortAscending = true;
                sortButton.setText("Sort by Last Name (Ascending)");
            }
            sortedEmployees.setComparator(comparator);
        });

        Button saveButton = new Button("Save Employees to File");
        saveButton.setOnAction(event -> {
            // Save employees to file
            FileChooser fileChooser = new FileChooser();
            fileChooser.setTitle("Save Employees to File");
            File file = fileChooser.showSaveDialog(primaryStage);
            if (file != null) {
                try (ObjectOutputStream out = new ObjectOutputStream(new FileOutputStream(file))) {
                    out.writeObject(new ArrayList<>(employees));
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        });

        Button loadButton = new Button("Load Employees from File");
        loadButton.setOnAction(event -> {
            // Load employees from file
            FileChooser fileChooser = new FileChooser();
            fileChooser.setTitle("Load Employees from File");
            File file = fileChooser.showOpenDialog(primaryStage);
            if (file != null) {
                try (ObjectInputStream in = new ObjectInputStream(new FileInputStream(file))) {
                    employees.setAll((List<Employee>) in.readObject());
                } catch (IOException | ClassNotFoundException e) {
                    e.printStackTrace();
                }
            }
        });

        ToggleGroup presentToggleGroup = new ToggleGroup();
        RadioButton allRadioButton = new RadioButton("All");
        allRadioButton.setToggleGroup(presentToggleGroup);
        allRadioButton.setSelected(true);
        RadioButton presentRadioButton = new RadioButton("Present");
        presentRadioButton.setToggleGroup(presentToggleGroup);
        RadioButton notPresentRadioButton = new RadioButton("Not Present");
        notPresentRadioButton.setToggleGroup(presentToggleGroup);

        presentToggleGroup.selectedToggleProperty().addListener((observable, oldValue, newValue) -> {
            // Update filtered employees based on selected toggle
            if (newValue == allRadioButton) {
                filteredEmployees.setPredicate(employee -> true);
            } else if (newValue == presentRadioButton) {
                filteredEmployees.setPredicate(Employee::isPresent);
            } else if (newValue == notPresentRadioButton) {
                filteredEmployees.setPredicate(employee -> !employee.isPresent());
            }
        });

        HBox dataHBox = new HBox(10, loadButton, saveButton);
        dataHBox.setSpacing(10);
        dataHBox.setAlignment(Pos.CENTER);
        dataHBox.setPadding(new Insets(10));
        
        HBox presentHBox = new HBox(10, allRadioButton, presentRadioButton, notPresentRadioButton, sortButton); presentHBox.setSpacing(10);
        presentHBox.setPadding(new Insets(10));
        presentHBox.setAlignment(Pos.CENTER);
        
        VBox tableVBox = new VBox(searchField, presentHBox, employeeListView);
        
        VBox newRecordVBox = new VBox(lastNameField, positionField, departmentField, badgeNumberField, isPresentCheckBox, lastEntryExitTimeField, addButton);
        newRecordVBox.setSpacing(10);
        newRecordVBox.setAlignment(Pos.CENTER);
        newRecordVBox.setPadding(new Insets(10));

        // Create layout and scene
        BorderPane root = new BorderPane();
        root.setBottom(dataHBox);
        root.setLeft(tableVBox);
        root.setRight(newRecordVBox);
        root.setPadding(new Insets(10));
        Scene scene = new Scene(root);

        // Set up stage
        primaryStage.setScene(scene);
        primaryStage.setTitle("Employee List");
        primaryStage.show();
    }

    public static void main(String[] args) {
        launch(args);
    }
}

File: L6/EmployeeManagement (Kotlin)/src/main/kotlin/com/example/employeemanagement/EmployeeManagement.kt
package com.example.employeemanagement

import javafx.application.Application
import javafx.beans.value.ObservableValue
import javafx.collections.FXCollections
import javafx.collections.transformation.FilteredList
import javafx.collections.transformation.SortedList
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.geometry.Insets
import javafx.geometry.Pos
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.layout.BorderPane
import javafx.scene.layout.HBox
import javafx.scene.layout.VBox
import javafx.stage.FileChooser
import javafx.stage.Stage
import java.io.*
import java.util.*

class EmployeeManagement : Application() {
    private val employees = FXCollections.observableArrayList<Employee>()
    private val filteredEmployees = FilteredList(employees)
    private val sortedEmployees = SortedList(filteredEmployees)
    private var sortAscending = true
    override fun start(primaryStage: Stage) {
        // Create UI elements
        val employeeListView = ListView(sortedEmployees)
        val searchField = TextField()
        searchField.promptText = "Search by Last Name"
        searchField.textProperty().addListener { observable: ObservableValue<out String?>?, oldValue: String?, newValue: String? ->
            // Update filtered employees based on search text
            filteredEmployees.setPredicate { employee: Employee ->
                if (newValue == null || newValue.isEmpty()) {
                    return@setPredicate true
                }
                val lowerCaseFilter = newValue.lowercase(Locale.getDefault())
                employee.lastName.lowercase(Locale.getDefault()).contains(lowerCaseFilter)
            }
        }
        val lastNameField = TextField()
        lastNameField.promptText = "Last Name"
        val positionField = TextField()
        positionField.promptText = "Position"
        val departmentField = TextField()
        departmentField.promptText = "Department"
        val badgeNumberField = TextField()
        badgeNumberField.promptText = "Badge Number"
        val isPresentCheckBox = CheckBox("Is Present")
        val lastEntryExitTimeField = TextField()
        lastEntryExitTimeField.promptText = "Last Entry/Exit Time"
        val contextMenu = ContextMenu()
        val removeItem = MenuItem("Remove")
        removeItem.onAction = EventHandler { e: ActionEvent? ->
            // Remove selected employee from list
            val selectedEmployee = employeeListView.selectionModel.selectedItem
            employees.remove(selectedEmployee)
        }
        contextMenu.items.add(removeItem)
        employeeListView.contextMenu = contextMenu
        val addButton = Button("Add Employee")
        addButton.onAction = EventHandler<ActionEvent> { event: ActionEvent? ->
            // Add employee to list
            val lastName = lastNameField.text
            val position = positionField.text
            val department = departmentField.text
            val badgeNumber = badgeNumberField.text.toInt()
            val isPresent = isPresentCheckBox.isSelected
            val lastEntryExitTime = lastEntryExitTimeField.text
            val employee = Employee(lastName, position, department, badgeNumber, isPresent, lastEntryExitTime)
            employees.add(employee)

            // Clear input fields
            lastNameField.clear()
            positionField.clear()
            departmentField.clear()
            badgeNumberField.clear()
            isPresentCheckBox.isSelected = false
            lastEntryExitTimeField.clear()
        }
        val saveButton = Button("Save Employees to File")
        saveButton.onAction = EventHandler { event: ActionEvent? ->
            // Save employees to file
            val fileChooser = FileChooser()
            fileChooser.title = "Save Employees to File"
            val file = fileChooser.showSaveDialog(primaryStage)
            if (file != null) {
                try {
                    ObjectOutputStream(FileOutputStream(file)).use { out -> out.writeObject(ArrayList(employees)) }
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
        }
        val loadButton = Button("Load Employees from File")
        loadButton.onAction = EventHandler { event: ActionEvent? ->
            // Load employees from file
            val fileChooser = FileChooser()
            fileChooser.title = "Load Employees from File"
            val file = fileChooser.showOpenDialog(primaryStage)
            if (file != null) {
                try {
                    ObjectInputStream(FileInputStream(file)).use { `in` -> employees.setAll(`in`.readObject() as List<Employee>) }
                } catch (e: IOException) {
                    e.printStackTrace()
                } catch (e: ClassNotFoundException) {
                    e.printStackTrace()
                }
            }
        }
        val sortButton = Button("Sort by Last Name")
        sortButton.onAction = EventHandler { event: ActionEvent? ->
            // Sort employees by last name
            val comparator: Comparator<Employee>
            if (sortAscending) {
                comparator = Comparator.comparing(Employee::lastName)
                sortAscending = false
                sortButton.text = "Sort by Last Name (Descending)"
            } else {
                comparator = Comparator.comparing(Employee::lastName).reversed()
                sortAscending = true
                sortButton.text = "Sort by Last Name (Ascending)"
            }
            sortedEmployees.setComparator(comparator)
        }
        val presentToggleGroup = ToggleGroup()
        val allRadioButton = RadioButton("All")
        allRadioButton.toggleGroup = presentToggleGroup
        allRadioButton.isSelected = true
        val presentRadioButton = RadioButton("Present")
        presentRadioButton.toggleGroup = presentToggleGroup
        val notPresentRadioButton = RadioButton("Not Present")
        notPresentRadioButton.toggleGroup = presentToggleGroup
        presentToggleGroup.selectedToggleProperty().addListener { observable: ObservableValue<out Toggle>?, oldValue: Toggle?, newValue: Toggle ->
            // Update filtered employees based on selected toggle
            if (newValue === allRadioButton) {
                filteredEmployees.setPredicate { employee: Employee? -> true }
            } else if (newValue === presentRadioButton) {
                filteredEmployees.setPredicate { employee: Employee -> employee.isPresent }
            } else if (newValue === notPresentRadioButton) {
                filteredEmployees.setPredicate { employee: Employee -> !employee.isPresent }
            }
        }
        val dataHBox = HBox(10.0, loadButton, saveButton)
        dataHBox.spacing = 10.0
        dataHBox.alignment = Pos.CENTER
        dataHBox.padding = Insets(10.0)
        val presentHBox = HBox(10.0, allRadioButton, presentRadioButton, notPresentRadioButton, sortButton)
        presentHBox.spacing = 10.0
        presentHBox.padding = Insets(10.0)
        presentHBox.alignment = Pos.CENTER
        val tableVBox = VBox(searchField, presentHBox, employeeListView)
        val newRecordVBox = VBox(lastNameField, positionField, departmentField, badgeNumberField, isPresentCheckBox, lastEntryExitTimeField, addButton)
        newRecordVBox.spacing = 10.0
        newRecordVBox.alignment = Pos.CENTER
        newRecordVBox.padding = Insets(10.0)

        // Create layout and scene
        val root = BorderPane()
        root.bottom = dataHBox
        root.left = tableVBox
        root.right = newRecordVBox
        root.padding = Insets(10.0)
        val scene = Scene(root)

        // Set up stage
        primaryStage.scene = scene
        primaryStage.title = "Employee List"
        primaryStage.show()
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            launch(EmployeeManagement::class.java)
        }
    }
}
